lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,283 @@
+    } else {
+      toggleExpandButton.addEventListener('click', () => toggleExpandedFn());
+    }
+
+    const nodeContainer = dom.find('.lh-code-snippet__node', header);
+    if (details.node && dom.isDevTools()) {
+      nodeContainer.appendChild(detailsRenderer.renderNode(details.node));
+    }
+
+    return header;
+  }
+
+  /**
+   * Renders a line of DOM content (code, message, or empty line)
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineDetails} lineDetails
+   * @return {Element}
+   */
+  static renderSnippetLine(dom, tmpl, {content, number, truncated, contentType, visibility}) {
+    const clonedTemplate = dom.cloneTemplate('#tmpl-lh-code-snippet__line', tmpl);
+    const codeLine = dom.find('.lh-code-snippet__line', clonedTemplate);
+    const {classList} = codeLine;
+
+    const typeClasses = {
+      [LineContentType.CODE_NORMAL]: ['lh-code-snippet__line--code'],
+      [LineContentType.CODE_HIGHLIGHTED]: [
+        'lh-code-snippet__line--code',
+        'lh-code-snippet__line--highlighted',
+      ],
+      [LineContentType.EMPTY]: ['lh-code-snippet__line--empty'],
+      [LineContentType.MESSAGE]: ['lh-code-snippet__line--message'],
+    }[contentType];
+    typeClasses.forEach(typeClass => classList.add(typeClass));
+
+    if (visibility === LineVisibility.WHEN_COLLAPSED) {
+      classList.add('lh-code-snippet__show-if-collapsed');
+    }
+    if (visibility === LineVisibility.WHEN_EXPANDED) {
+      classList.add('lh-code-snippet__show-if-expanded');
+    }
+
+    const lineContent = content + (truncated ? '…' : '');
+    const lineContentEl = dom.find('.lh-code-snippet__line code', codeLine);
+    if (contentType === LineContentType.MESSAGE) {
+      lineContentEl.appendChild(dom.convertMarkdownLinkSnippets(lineContent));
+    } else {
+      lineContentEl.textContent = lineContent;
+    }
+
+    dom.find('.lh-code-snippet__line-number', codeLine).textContent = number.toString();
+
+    return codeLine;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippetHighlight} highlight
+   * @return {Element}
+   */
+  static renderHighlightMessage(dom, tmpl, highlight) {
+    return CodeSnippetRenderer.renderSnippetLine(dom, tmpl, {
+      number: ' ',
+      content: highlight.message,
+      contentType: LineContentType.MESSAGE,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LineVisibility} visibility
+   * @return {Element}
+   */
+  static renderOmittedLines(dom, tmpl, visibility) {
+    return CodeSnippetRenderer.renderSnippetLine(dom, tmpl, {
+      number: '…',
+      content: '',
+      visibility,
+      contentType: LineContentType.EMPTY,
+    });
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
```

Comment:
I think some explanation of the overall structure would help a bit here.

something like the below if it's accurate? please feel free to adjust to what's happening :)

- First we add the highlights that relate to the whole snippet and don't reference particular lines
- Then we'll render the lines with each of their line-specific highlights and omitted segments
- Finally we'll make sure we render the omitted lines placeholder for the beginning and end